lib/cretonne/src/verifier.rs

```diff
@@ -284,11 +284,34 @@ impl<'a> Verifier<'a> {
     }
 
     fn verify_value(&self, inst: Inst, v: Value) -> Result<()> {
-        if !self.func.dfg.value_is_valid(v) {
-            err!(inst, "invalid value reference {}", v)
-        } else {
-            Ok(())
+        let dfg = &self.func.dfg;
+        if !dfg.value_is_valid(v) {
+            return err!(inst, "invalid value reference {}", v);
         }
+
+        // SSA form
+        match dfg.value_def(v) {
+            // Value is defined by an instruction that exists and is inserted in an EBB.
+            ValueDef::Res(inst, _) => {
+                if !dfg.insts.is_valid(inst) {
+                    return err!(v, "value is defined by invalid instruction {}", inst);
```

Comment:
The entity the you pass as the first argument to `err!` works like a source location. When possible, we will even use the parser's source map to convert it to a real line number.

So here it would be better to go:

```rust
            ValueDef::Res(def_inst, _) => {
                if !dfg.insts.is_valid(def_inst) {
                    return err!(inst, "{} is defined by invalid instruction {}", v, inst);
```

We won't be able to map an invalidly defined value to a proper source location.

Other times, using a value as a source location is just fine.